### PR TITLE
Support status reporting across multiple OpenStack regions

### PR DESCRIPTION
**Pull request by neiljerram**
## Release Note

```release-note
To do
```

### CODE REVIEW

File: lib/backend/model/statusreports.go [hunk 1]
```diff
@@ -59,10 +60,11 @@ func (key ActiveStatusReportKey) String() string {
 
 type ActiveStatusReportListOptions struct {
 	Hostname string
+	Region   string
 }
 
 func (options ActiveStatusReportListOptions) defaultPathRoot() string {
-	k := "/calico/felix/v1/host"
+	k := "/calico/felix/v2/" + regionString(options.Region) + "/host"
```
Comment (fasaxc) [inline lib/backend/model/statusreports.go]: As discussed, code does what felix needs but it's breaking the normal convention in this API (where empty string means "don't filter on this field").

Comment (neiljerram) [reply lib/backend/model/statusreports.go]: Fixed in next commit by changing the `Region` fields to `RegionString`.  A region _string_ is never validly empty.

File: lib/backend/model/statusreports.go [hunk 2]
```diff
@@ -24,13 +24,14 @@ import (
...
-	matchLastStatusReport   = regexp.MustCompile("^/?calico/felix/v1/host/([^/]+)/last_reported_status")
+	matchActiveStatusReport = regexp.MustCompile("^/?calico/felix/v2/([^/]+)/host/([^/]+)/status$")
+	matchLastStatusReport   = regexp.MustCompile("^/?calico/felix/v2/([^/]+)/host/([^/]+)/last_reported_status")
 	typeStatusReport        = reflect.TypeOf(StatusReport{})
 )
 
 type ActiveStatusReportKey struct {
-	Hostname string `json:"-" validate:"required,hostname"`
+	Hostname     string `json:"-" validate:"required,hostname"`
+	RegionString string
```
Comment (fasaxc) [inline lib/backend/model/statusreports.go]: Missing the json/validate tags?

File: lib/backend/model/statusreports.go [hunk 3]
```diff
@@ -77,16 +82,22 @@ func (options ActiveStatusReportListOptions) KeyFromDefaultPath(ekey string) Key
...
 		return nil
 	}
-	return ActiveStatusReportKey{Hostname: name}
+	return ActiveStatusReportKey{Hostname: name, RegionString: regionString}
 }
 
 type LastStatusReportKey struct {
-	Hostname string `json:"-" validate:"required,hostname"`
+	Hostname     string `json:"-" validate:"required,hostname"`
+	RegionString string
```
Comment (fasaxc) [inline lib/backend/model/statusreports.go]: Missing tags?

File: lib/backend/model/workloadendpointstatus.go [hunk 4]
```diff
@@ -87,10 +93,11 @@ type WorkloadEndpointStatusListOptions struct {
 	OrchestratorID string
 	WorkloadID     string
 	EndpointID     string
+	RegionString   string
 }
 
 func (options WorkloadEndpointStatusListOptions) defaultPathRoot() string {
-	k := "/calico/felix/v1/host"
+	k := "/calico/felix/v2/" + options.RegionString + "/host"
```
Comment (fasaxc) [inline lib/backend/model/workloadendpointstatus.go]: I think if the region string is empty, this should return `"/calico/felix/v2/"`

File: lib/backend/model/statusreports.go (hunk 2 above)
Comment (fasaxc) [reply lib/backend/model/statusreports.go]: We should probably validate that it doesn't have a `/` in it if nothing else.

Comment (fasaxc) [reply lib/backend/model/statusreports.go]: Just a thought: could you make RegionString into a typedef?  I.e. `type RegionString string`  Then the user would have to go through the RegionString function.

File: lib/backend/model/workloadendpointstatus.go [hunk 5]
```diff
@@ -26,14 +26,15 @@ import (
...
-	matchWorkloadEndpointStatus = regexp.MustCompile("^/?calico/felix/v1/host/([^/]+)/workload/([^/]+)/([^/]+)/endpoint/([^/]+)$")
+	matchWorkloadEndpointStatus = regexp.MustCompile("^/?calico/felix/v2/([^/]+)/host/([^/]+)/workload/([^/]+)/([^/]+)/endpoint/([^/]+)$")
 )
 
 type WorkloadEndpointStatusKey struct {
 	Hostname       string `json:"-"`
 	OrchestratorID string `json:"-"`
 	WorkloadID     string `json:"-"`
 	EndpointID     string `json:"-"`
+	RegionString   string
```
[the remark this answers is not in the document]
Comment (neiljerram) [reply lib/backend/model/workloadendpointstatus.go]: I don't know.  I suspect it was just cargo culting, and thought that any real need should be revealed by tests, and so left it off the new field here.

(The json and validate tags are used for validating some v1 structures - https://github.com/projectcalico/libcalico-go/blob/master/lib/validator/v1/validator.go#L91 - but that never applies to Key and Options structures.)

File: lib/backend/model/statusreports.go (hunk 2 above)
Comment (neiljerram) [reply lib/backend/model/statusreports.go]: Yes, that's a good idea, and I could make the RegionString function do escaping.

File: lib/backend/model/statusreports.go (hunk 3 above)
Comment (neiljerram) [reply lib/backend/model/statusreports.go]: (As above)

File: lib/backend/model/statusreports.go (hunk 2 above)
Comment (neiljerram) [reply lib/backend/model/statusreports.go]: Actually no, that requires too much change everywhere, and is inconsistent which all the other string-like fields that are just 'string'.  Let's just do the validation.

File: lib/backend/model/workloadendpointstatus.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fasaxc) [reply lib/backend/model/workloadendpointstatus.go]: > As a special case, if the field tag is "-", the field is always omitted. Note that a field with name "-" can still be generated using the tag "-,".

So, why are we omitting those fields from JSON?

[the remark this answers is not in the document]
Comment (fasaxc) [reply lib/backend/model/workloadendpointstatus.go]: https://golang.org/pkg/encoding/json/

File: lib/backend/model/statusreports.go (hunk 2 above)
Comment (fasaxc) [reply lib/backend/model/statusreports.go]: OK, can live with that; should there be a `validate` tag, then?

File: lib/backend/model/statusreports.go (hunk 3 above)
Comment (fasaxc) [reply lib/backend/model/statusreports.go]: I don't follow, shouldn't this have a validate tag even if the json tag isn't relevant?

File: lib/backend/model/statusreports.go [hunk 6]
```diff
@@ -97,7 +116,13 @@ func (key LastStatusReportKey) defaultDeletePath() (string, error) {
 	if key.Hostname == "" {
 		return "", errors.ErrorInsufficientIdentifiers{Name: "hostname"}
 	}
-	e := fmt.Sprintf("/calico/felix/v1/host/%s/last_reported_status", key.Hostname)
+	if key.RegionString == "" {
+		return "", errors.ErrorInsufficientIdentifiers{Name: "regionString"}
+	}
+	if strings.Contains(key.RegionString, "/") {
```
Comment (fasaxc) [inline lib/backend/model/statusreports.go]: Why implement this validation differently from every other key, which seems to use the validate tag?

**Pull request merged**